python/fapolicy_analyzer/ui/state_manager.py

```diff
@@ -83,17 +88,82 @@ def __update_dirty_queue(self):
 
-    def add_system_notification(
-        self, notification: str, notification_type: NotificationType
-    ):
+    def __path_action_list_2_queue(self, listPathAction):
+        """Converts a list of Path/Action tuples to populate changeset queue
+This is a utility function intended for unit testing and troubleshooting
+the StateManager and its interactions with interfacing objects. It is used to
+populate the internal queue structure with Changeset objects, however it does
+not exercise the full normal end to end data path which will normally apply and
+deploy these changesets.
+"""
+        for e in listPathAction:
+            cs = Changeset()
+            if e[1] == "Add":
+                cs.add_trust(e[0])
+            elif e[1] == "Del":
+                cs.del_trust(e[0])
+            else:
+                print("Error: Path/Action Pair: "
+                      "Unknown Action: {} {}".format(e[0], e[1]))
+                continue
+            self.add_changeset_q(cs)
+
+    def __path_action_list_to_dict(self):
+        """Convert Path/Action list of tuple pairs to dict for json xfer"""
+        dictPA = dict()
+        for t in self.get_path_action_list():
+            dictPA[t[0]] = t[1]
+        logging.debug("Path/Action Dict: {}".format(dictPA))
+        return dictPA
+
+    def __path_action_dict_to_list(self, dictPathAction):
+        """Convert Path/Action dict to list of tuple pairs for json xfer"""
+        listPaTuples = list()
+        for e in dictPathAction.keys():
+            listPaTuples.append((e, dictPathAction[e]))
+        logging.debug("PathAction Tuple List: {}".format(listPaTuples))
+        return listPaTuples
+
+    # ######################## Edit Session Mgmt ############################
+    def save_edit_session(self, strJsonFile):
+        """Save the pending changeset queue to the specified json file"""
+        logging.debug("StateManager::save_edit_session({})".format(strJsonFile))
+        with open(strJsonFile, "w") as fp:
+            json.dump(self.__path_action_list_to_dict(), fp,
+                      sort_keys=True,
+                      indent=4)
+
+    def open_edit_session(self, strJsonFile):
+        """Save the FIFO changeset queue to the specified json file on disk"""
+        logging.debug("Entered StateManager::open_edit_session({})"
+                      .format(strJsonFile))
+        listPA = None
+        with open(strJsonFile, "r") as fp:
+            d = json.load(fp, object_pairs_hook=OrderedDict)
+            logging.debug("Loaded dict = ", d)
+            listPA = self.__path_action_dict_to_list(d)
+            logging.debug("StateManager::open_edit_session():{}".format(listPA))
+
+            # Deleting current edit session history prior to replacing it.
+            if listPA:
+                # ToDo: Delete pending ops in the ATDA's embedded TreeView
+                self.del_changeset_q()
```

Comment:
This doesn't delete the changesets from the rust side.  Is that coming in a later commit?

Reply:
No. The ATDA should be the primary client of the rust fapolicyd layer IMO. It also allows it to maintain its UI components so they are consistent with the databases.
Similar, to how the StateManager leverages the ATDA's existing code to add Changesets via an event, I'd like to see the ATDA's api extended so the StateManager can request it to delete the existing undeployed 
ops. Whether the ATDA should contain a queue itself, or just keep things in the existing TreeStore is TBD. I'd like to know your thoughts.